content/post/2019-10-15.Rmd

```diff
@@ -0,0 +1,388 @@
+library(purrrlyr)
+
+n <- 5
+
+# Initialization of our data frame
+df <- tibble(x = numeric(n+1),
+             y = numeric(n+1))
+
+# Convert our data frame into a list
+by_row(df, function(v) list(v)[[1L]], .collate = "list")$.out -> df
+
+# This function computes current location dependind of previous one
+f <- function(j, k, a, b, c, d) {
+  tibble(
+    x = sin(a*j$y)+c*cos(a*j$x),
+    y = sin(b*j$x)+d*cos(b*j$y)
+  )
+}
+
+# We apply accumulate on our list to store all steps and convert into a data frame
+accumulate(df, f, a, b, c, d) %>% map_df(~ .x) -> df
+
+df
+```
+
+Even this approach is not efficient since the `accumulate` function must be applied on an existing object (the `df` in our case)
+
+## Combining C++ and R: the good way
+
+The next paragraph, extracted from Hadley Wickham's [Advanced R](https://adv-r.hadley.nz/rcpp.html), explains the motivations of combining C++ and R:
+
+`Rcpp` makes it very simple to connect C++ to R and provides a clean, approachable API that lets you write high-performance code, insulated from R’s complex C API. Typical bottlenecks that C++ can address include:
+
++ Loops that can’t be easily vectorised because subsequent iterations depend on previous ones.
+
++ Recursive functions, or problems which involve calling functions millions of times. The overhead of calling a function in C++ is much lower than in R.
+
++ Problems that require advanced data structures and algorithms that R doesn’t provide. Through the standard template library (STL), C++ has efficient implementations of many important data structures, from ordered maps to double-ended queues.
+
+This code populates our data frame using C++:
+
+```{r, warning = FALSE, message = FALSE}
+
+# Load in libaries
+library(Rcpp)
+library(tidyverse)
+
+# Here comes the C++ code
+cppFunction('DataFrame createTrajectory(int n, double x0, double y0,
+            double a, double b, double c, double d) {
+            // create the columns
+            NumericVector x(n);
+            NumericVector y(n);
+            x[0]=x0;
+            y[0]=y0;
+            for(int i = 1; i < n; ++i) {
+            x[i] = sin(a*y[i-1])+c*cos(a*x[i-1]);
+            y[i] = sin(b*x[i-1])+d*cos(b*y[i-1]);
+            }
+            // return a new data frame
+            return DataFrame::create(_["x"]= x, _["y"]= y);
+            }
+            ')
+
+a <- -1.25
+b <- -1.25
+c <- -1.82
+d <- -1.91
+
+# Call the function starting at (0,0) to store 10M jumps
+df <- createTrajectory(10000000, 0, 0, a, b, c, d)
+
+head(df)
+```
+
+Note that, apart of `Rcpp`,  we also loaded the tidyverse package. This is becaus we will use some of its features, as `dplyr` or `ggplot2`, later. 
+
+The function `createTrajectory` is written in `C++` and stores the itinerary of a initial point. You can try it to see that It's extremely fast. In the previous example, I populated on the fly our data frame with 10 milion points starting at `(0,0)` in less than 5 seconds in a conventional laptop.
```

Comment:
1/... itinerary of AN initial point. Sería con AN en vez de A.

2/ En la palabra million, falta una 'L'.